wordpress-mobile/WordPress-Android
# Migrate QRCode Auth to Compose

Fixes #16933

This PR Refactors the UI of the QRCode Auth screen using Jetpack Compose.

Since it's the first UI built with Compose in the project, there are a few other changes added:
- Adapt `detekt.yml` according to the [Configuration for Compose](https://detekt.dev/docs/introduction/compose/) guide.
- Add dependency on compose `ui-tooling` to enable working with `@Preview`'s for composable functions
- Add dependency on `lifecycle-viewmodel-compose` to facilite communicating with the `viewModel` in composables
- Add `uiStringText` util composable https://github.com/wordpress-mobile/WordPress-Android/commit/29a5f93b2e45ba74895302906ea330ff22c5aaf8
- Adapt the Compose theme introduced in [PR 16206](https://github.com/wordpress-mobile/WordPress-Android/pull/16206) with custom colors for the Jetpack and WordPress apps
- Add other basic building blocks for working with Compose (theme, fontSizes, dimens, etc)

To test:
1. Open the Jetpack App from this PR or clone the branch and run the `jetpackWasabiDebug` app flavor
2. Go to `Me` → `Scan Log-in Code`
3. On the laptop in a browser open a guest session and go to
   https://wordpress.com/log-in/qr?redirect_to=https%3A%2F%2Fwordpress.com%2F
4. Scan the QR Code with the mobile app
   **expect** a success screen (see 'Validated' screenshot)
5. Tap `Yes, log me in`
   **expect** an authenticating screen (see 'Authenticated' screenshot) then
   **expect** a done screen (see 'Done' screenshot)
6. Tap `Dismiss` and Go back to `Me` → `Scan Log-in Code`
7. Open another guest session in your browser on the laptop and go to
   https://apps.wordpress.com/mobile/#qr-code-login?token=asdfadsfa&data=asdfasdf
8. Scan the QR Code with the mobile app
   **expect** an error screen (see 'Error • Invalid' screenshot)
9. Additionally you can test with the internet connection disabled on the device to scan a valid QR code (see step 3)
   **expecting** an error screen (see 'Error • Network' screenshot)

For more context about the migrated feature check out the parent issue that introduced it: #16481.

⚠️ This PR is ready to merge only when the testing confirms the feature still works and looks like the initial implementation.

**Previews:**

| Validated | Authenticating | Done | 
| --- | --- | --- |
| <img width="250" src="https://user-images.githubusercontent.com/4588074/180257362-514079c8-55eb-41e4-b8a5-071f7b74f929.png"> | <img width="250" src="https://user-images.githubusercontent.com/4588074/180267277-f8fe3bd3-3e7a-42c1-96ce-961b393c82cf.png"> | <img width="250" src="https://user-images.githubusercontent.com/4588074/180257358-b581747f-330f-4393-a08f-4ee6c8d6e62c.png"> |

| Error • Invalid | Error • Network |
| --- | --- |
| <img width="250" src="https://user-images.githubusercontent.com/4588074/180257352-8d7755bd-6861-4115-9f06-2d1af69c4b1b.png"> | <img width="250" src="https://user-images.githubusercontent.com/4588074/180257347-4ecf3fba-ee3f-4793-9825-45d0b41510ea.png"> |

## Regression Notes
1. Potential unintended areas of impact
   Functionality on the QR Code Auth feature.

6. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual Testing.

7. What automated tests I added (or what prevented me from doing so)
   Relied on the existing unit tests. `Compose` UI testing is nice but in order to promote the Compose adoption we're not going to tackle it in this PR.

PR submission checklist:
